modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TableViewTableHeaderRowTest.java

```diff
@@ -225,6 +228,70 @@ void testOverriddenColumnPopupMenu() {
         assertNull(columnPopupMenu);
     }
 
+    /**
+     * Tests that re-setting the same columns does not cause memory leaks.
+     * See also: <a href="https://bugs.openjdk.org/browse/JDK-8341687">JDK-8341687</a>.
+     */
+    @Test
+    void testReSettingColumnsDoesNotLeakMemory() {
+        for (TableColumn<String, ?> column : tableView.getColumns()) {
+            assertEquals(2, getVisibilityListenerCount(column));
+        }
+
+        // Trigger the menu once so that it will start listening to column changes.
+        MouseEventFirer mouseEventFirer = new MouseEventFirer(cornerRegion);
+        mouseEventFirer.fireMousePressed();
+
+        // Now the table menu is listening for changes as well.
+        for (TableColumn<String, ?> column : tableView.getColumns()) {
+            assertEquals(3, getVisibilityListenerCount(column));
+        }
+
+        tableView.getColumns().setAll(FXCollections.observableArrayList(tableView.getColumns()));
+        tableView.getColumns().setAll(FXCollections.observableArrayList(tableView.getColumns()));
+
+        // The count should be the same still.
+        for (TableColumn<String, ?> column : tableView.getColumns()) {
+            assertEquals(3, getVisibilityListenerCount(column));
+        }
+    }
+
+    /**
+     * Tests that toggling the column visibility does not cause memory leaks.
+     * See also: <a href="https://bugs.openjdk.org/browse/JDK-8341687">JDK-8341687</a>.
```

Comment:
same comment - does not test for memory leak.  the memory leak test might involve `JMemoryBuddy` - see for example `SystemMenuBarTest`.

Reply:
I tried, but it isn't that easy - since the memory is not leaked because we keep some references we should not keep - rather we are adding listener again and again and again. So I did not managed to make a test with `JMemoryBuddy`, but can prove the point with the amount of listener instead.

Reply:
I think this is an acceptable explanation as it tests the root cause of the leak.

Do you think changing the skin would allow us to create a test for the memory leak specifically.  Maybe some variation of `SkinMemoryLeakTest` ?

Reply:
Found a way to test it with `JMemoryBuddy`. Kept the listener tests as well, since they also prove the point.